components/experimental/src/measure/measureunit.rs

```diff
@@ -62,160 +61,90 @@ impl MeasureUnit {
-                .iter()
-                .enumerate()
-                .fold(String::new(), |acc, (index, unit)| {
-                    if index < negative_prefix_units.len() - 1 {
-                        acc + &unit.to_string() + "#"
-                    } else {
-                        acc + &unit.to_string()
-                    }
-                });
-
-        // add constant part to the negative part
+        let mut short_representation = String::new();
+
         if self.constant_denominator > 0 {
-            if !negative_part.is_empty() {
-                negative_part.insert(0, '#');
-            }
-            negative_part.insert_str(0, &power_of_10_to_scientific(self.constant_denominator));
-            negative_part.insert(0, 'C');
+            short_representation.push('C');
+            append_power_of_10_to_scientific(self.constant_denominator, &mut short_representation);
         }
 
-        // Glue the positive and negative parts together to get the final short representation
-        let mut short_representation = positive_part;
-        if !short_representation.is_empty() && !negative_part.is_empty() {
-            short_representation.push_str("#R#"); // --> -per- representation
-        } else if !negative_part.is_empty() {
-            short_representation.push_str("R#"); // --> per- representation
-        }
+        self.single_units.as_slice().iter().for_each(|single_unit| {
+            single_unit.append_short_representation(&mut short_representation)
+        });
 
-        short_representation.push_str(&negative_part);
         short_representation
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use icu::experimental::measure::parser::MeasureUnitParser;
+
+    #[test]
+    fn test_generate_short_representation() {
+        let parser = MeasureUnitParser::new();
+
+        let test_cases = vec![
+            ("meter", "I85", 2),
+            ("foot", "I50", 1),
+            ("inch", "I66", 1),
+            ("square-meter", "P2I85", 7),
+            ("square-millimeter", "P2D-3I85", 9),
+            ("micrometer", "D-6I85", 4),
+            ("meter-per-second", "I85P-1I127", 6),
+            ("liter-per-100-kilometer", "C100I82P-1D3I85", 8),
+            ("portion-per-1e9", "C1E9I113", 7),
+            ("per-10000000000-portion", "C1E10P-1I113", 11),
+            ("liter-per-240000000000-kilometer", "C24E10I82P-1D3I85", 15),
+            ("millimeter-per-square-microsecond", "D-3I85P-2D-6I127", 17),
+        ];
+
+        for (full_unit, expected_short, expected_diff) in test_cases {
+            let measure_unit = parser.try_from_str(full_unit).unwrap();
+            let short_representation = measure_unit.generate_short_representation();
+            assert_eq!(short_representation, expected_short, "{}", full_unit);
+            assert_eq!(
+                full_unit.len() - short_representation.len(),
+                expected_diff,
+                "{}",
+                full_unit
+            );
```

Comment:
still, this is pointless assertion. it's at most useful for a code reviewer to demonstrate that the strings are shorter, but I can see that. there's no point in calculating the diff numbers when new test cases are added, or actually executing this code

Reply:
okay, as you like :)